media/src/commonMain/kotlin/MediaPlayer.kt

```diff
@@ -0,0 +1,442 @@
+                is PlaybackState.Ended -> throw PlaybackError.PlaybackHasEnded
+                is PlaybackState.Error -> if (resetOnError) {
+                    playbackStateRepo.takeAndChangeState(PlaybackState.Active::class) { it.reset }
+                } else throw state.error
+                is PlaybackState.Active -> playbackStateRepo.takeAndChangeState(PlaybackState.Active::class) { it.reset }
+            }
+        }.first()
+    }
+
+    override suspend fun renderVideoOnSurface(surface: MediaSurface?) {
+        playbackStateRepo.useState { state ->
+            when (state) {
+                is PlaybackState.Active -> state.mediaSurfaceController.renderVideoOnSurface(surface)
+                is PlaybackState.Ended -> throw PlaybackError.PlaybackHasEnded
+            }
+        }
+    }
+
+    private suspend fun start(playbackParameters: PlaybackState.PlaybackParameters) = try {
+        forceStart(playbackParameters, false)
+    } catch (e: PlaybackError) {
+        // Do nothing as this method should be safe to call
+    }
+
+    override suspend fun forceStart(playbackParameters: PlaybackState.PlaybackParameters, restartIfStarted: Boolean) = playbackStateRepo.transformLatest { state ->
+        when (state) {
+            is PlaybackState.Uninitialized -> {} // Do nothing until Initialized
+            is PlaybackState.Initialized -> {} // Do nothing until prepared
+            is PlaybackState.Idle -> playbackStateRepo.takeAndChangeState(PlaybackState.Idle::class) { it.play(playbackParameters) }
+            is PlaybackState.Completed -> playbackStateRepo.takeAndChangeState(PlaybackState.Completed::class) { it.start(playbackParameters) }
+            is PlaybackState.Stopped -> playbackStateRepo.takeAndChangeState(PlaybackState.Stopped::class) { it.reinitialize }
+            is PlaybackState.Playing -> {
+                if (state.playbackParameters != playbackParameters) {
+                    playbackStateRepo.takeAndChangeState(PlaybackState.Playing::class) {
+                        it.updatePlaybackParameters(playbackParameters)
+                    }
+                } else {
+                    if (restartIfStarted) {
+                        state.seekTo(Duration.ZERO)
+                    }
+                    emit(Unit)
+                }
+            }
+            is PlaybackState.Paused -> {
+                playbackStateRepo.takeAndChangeState(PlaybackState.Paused::class) {
+                    if (state.playbackParameters != playbackParameters) {
+                        it.updatePlaybackParameters(playbackParameters)
+                    } else {
+                        if (restartIfStarted) {
+                            state.seekTo(Duration.ZERO)
+                        }
+                        it.play
+                    }
+                }
+            }
+            is PlaybackState.Ended -> throw PlaybackError.PlaybackHasEnded
+            is PlaybackState.Error -> throw state.error
```

Comment:
I would also argue for Ended -> Closed, as ended first suggests to me it's the end of the video, until I remembed the `end` description in the README.

Exception in this case is the right call I think. We did discuss in the past to make a util method for this in test utils to encourage actually testing for it.